osd: allow EC PGs to do recovery below min_size

liuchang0812 opened the pull request:
Fixes: http://tracker.ceph.com/issues/18749

Signed-off-by: Chang Liu <liuchang@gmail.com>

Liuchang0812 commented:
@gregsfortytwo would you mind taking a look? thanks

Liuchang0812 commented:
jenkins test this please

liewegas commented:
@jdurgin @gregsfortytwo do you remember why we didn't allow this?

Liuchang0812 commented:
https://github.com/ceph/ceph/pull/8008  this patch updates min_size of EC pool to K+1.

gregsfortytwo commented:
The ticket has a pretty good explanation IIRC, and Sam said he was just paranoid.

Sorry for the lack of response @Liuchang0812; we've got a conference this week but this PR is on my list!

Liuchang0812 commented:
@gregsfortytwo ping

gregsfortytwo commented on src/osd/PG.cc:
```diff
@@ -1428,15 +1428,28 @@ bool PG::choose_acting(pg_shard_t &auth_log_shard_id,
-    dout(10) << "choose_acting failed, below min size" << dendl;
-    return false;
+  if (num_want_acting < pool.info.min_size) { 
+    bool ret = true;
```
The name ```ret``` is very confusing here. We return if ret is *false*, and if ret is true we just fall through and might or might not succeed later.

How about we call it "must_fail" or something else more descriptive?

gregsfortytwo commented on src/osd/PG.cc:
```diff
@@ -1428,15 +1428,28 @@ bool PG::choose_acting(pg_shard_t &auth_log_shard_id,
+      auto& ecp = get_osdmap()->get_erasure_code_profile(pool.info.erasure_code_profile);
+      auto pk = ecp.find("k");
+      assert(pk != ecp.end());
+      unsigned int k = atoi(pk->second.c_str());
```
I don't think this is quite right, since some of the erasure codes are quite odd.

Luckily, we already have a programmatic interface to tell if we have enough shards to recover a PG. It's the next check with the recoverable_predicate!

I think we just need to move this block below that check and then we don't need to do this check ourselves.

gregsfortytwo commented on src/osd/PG.cc:
```diff
@@ -1428,15 +1428,28 @@ bool PG::choose_acting(pg_shard_t &auth_log_shard_id,
 
-  // We go incomplete if below min_size for ec_pools since backfill
+  // We go incomplete if below data_chunk count for ec_pools since backfill
   // does not currently maintain rollbackability
```
Oh, and I think now we can just drop this comment. We'll only go incomplete if we are, in fact, incomplete. Which is the expected behavior.

gregsfortytwo commented:
I went through the uses of min_size and where we handle roll_forward and rollback everything looks good — PG::activate only sets last_epoch_started and last_interval_started if acting.size() > min_size; same for a roll forward. (And nobody else does a roll forward except for append_log, and that only when the OSD is a backfill recipient.)

Liuchang0812 commented:
@gregsfortytwo thank you for reviewing. I will update this PR based on your comments.

Liuchang0812 commented:
updated

gregsfortytwo commented on src/osd/PG.cc:
```diff
@@ -1429,23 +1429,19 @@ bool PG::choose_acting(pg_shard_t &auth_log_shard_id,
-    get_pgbackend()->get_is_recoverable_predicate());
-  if (!(*recoverable_predicate)(have)) {
+  if (!cct->_conf->osd_allow_recovery_below_min_size && 
+      pool.info.is_replicated() &&
```
While osd_allow_recovery_below_min_size definitely exists for debugging, there's no reason to restrict it to replicated pools.

Liuchang0812 replied:
Sorry for my late reply.  I tried to keep the code consistent with the previous. and I found that `osd_allow_recovery_below_min_size` is only used in here. maybe we could drop this option?

gregsfortytwo replied:
No, I mean that's an option mean to be used for testing and debugging. Replicated pools have various features that let them simulate an EC pool to deal with some of the ways in which peering is different between the two.

In this case, replicated pools have always been able to do recovery when below min_size, but we needed to be able to disable it so the behavior matches EC pools. Now that this PR is allowing recovery from any size, we may want this option to work on both pool types.

Although you're right; it becomes less useful once all pools can just recover whatever their size. Hrm.

gregsfortytwo commented:
This looks right at the basic level, but it's still pretty significant so I'm trying to figure out if we need to do anything in particular to validate it. @liewegas, do you think our regular thrashing is good enough? Or we could add tests which deliberately move around in the space between min_size and being recoverable...

Liuchang0812 commented:
@gregsfortytwo @tchaikov ping

Liuchang0812 commented:
I will rebase it, look close related code resolve gregs's comment this weekend. sorry.

Liuchang0812 commented:
Hey, is it possible that OSD is not recoverable even acting_set's size > min_size?  user wrongly set a min_size.

gregsfortytwo commented:
I'm not sure I buy that being good enough, @tchaikov. We want to validate that the recovery is actually happening, and that it doesn't break other parts of the system. Probabilistic tests are good for risks you can't identify but they aren't great for scenarios you know about.

Anyway, I merged this in with master and scheduled a 1/3000 subset run to see if it breaks any existing tests. Will try and spend some time in the teuthology thrashing code figuring out how to set up the specific scenarios this week.

tchaikov commented:
> Probabilistic tests are good for risks you can't identify but they aren't great for scenarios you know about.

agreed. would be nice to have a test covering this feature.

Liuchang0812 commented:
agreed. would be nice to have a test covering this feature.

I will try to write a test based on your comments as soon as possible. suggestions are appreciated. thank you

tchaikov commented:
@Liuchang0812 ping?

Liuchang0812 commented:
@Liuchang0812 ping?

sorry for late reply. tchaikov, gregsfortytwo. could you please give me some help.  what i should do is creating a teuthology task? are there any related code to be modified?

gregsfortytwo commented:
Yeah, I'm sorry I've been so quiet and didn't get this handled more promptly. :(

It's a few months old now but I actually rebased this patch at https://github.com/ceph/ceph-ci/commits/wip-allow-min-size-recovery and have a thoroughly-buggy min-size thrasher at https://github.com/gregsfortytwo/ceph/commits/wip-minsize-testing

IIRC the latest problem is that it is somehow not reading out the k and m minsize values when getting pool stats, so it's trying to kill the wrong number of OSDs at a time. Or maybe I'd gotten over that hump but the peering was still taking a long time and I needed to establish why?

It is perpetually the next thing on my list, but new stuff keeps getting added on top of it. I continue even now to think I'll get back to it late today or tomorrow.

Liuchang0812 commented:
@gregsfortytwo thanks for your kindly reply. I have rebased this PR. should I cherrpy-pick your commit, run teuthology test suite again?

gregsfortytwo commented:
I think it's just that we need code specifically for testing this condition, which doesn't exist outside of that still-buggy thrasher. If you think you can make that work, go for it! :)

Liuchang0812 commented:
@gregsfortytwo I updated theuthology thrasher task. I found that there was a `test_min_size` function already. so I updated `test_min_size` based on your new thrasher. would you mind taking a look? very thanks

I tried to deploy a in-house theuthology, but it's complicated. I need some more time to get theuthology running. so testing code in this PR is not tested.

neha-ojha commented:
@Liuchang0812 thanks for updating the PR, we will run this through teuthology testing and will share the results here.

neha-ojha commented on qa/tasks/ceph_manager.py:
```diff
@@ -659,27 +659,97 @@ def fix_pgp_num(self, pool=None):
+            k = 0
+            m = 99
+            has_pools = False
+            pools_json = self.get_osd_dump_json()['pools']:
```
`SyntaxError: invalid syntax` errors seen due to this

neha-ojha commented:
Test results: http://pulpito.ceph.com/nojha-2018-10-18_14:18:13-rados-wip-17619-distro-basic-smithi/

neha-ojha commented:
New set of test results: http://pulpito.ceph.com/nojha-2018-10-19_19:58:51-rados-wip-17619-18-10-19-distro-basic-smithi/

Liuchang0812 commented:
thank you @neha-ojha working on those failures

neha-ojha commented:
@Liuchang0812 have you had a chance to look into those failures?

Liuchang0812 commented:
> @Liuchang0812 have you had a chance to look into those failures?

sorry, I lost this task. I think I will look into those failures this weekend.

Liuchang0812 commented:
@neha-ojha it seems that all logs are expired. I got 404-not-found when I tried to open it.  see [teuthology log](http://qa-proxy.ceph.com/teuthology/nojha-2018-10-19_19:58:51-rados-wip-17619-18-10-19-distro-basic-smithi/3161427/teuthology.log)

neha-ojha commented:
@Liuchang0812 looks like @tchaikov is rerunning your PR through testing, we'll post the latest test results

tchaikov commented:
http://pulpito.ceph.com/kchai-2019-04-01_10:38:29-rados-wip-kefu-testing-2019-04-01-1531-distro-basic-mira/

the test run is not finished yet. but some failures might be relevant.

tchaikov commented on qa/tasks/ceph_manager.py:
```diff
@@ -659,27 +659,97 @@ def fix_pgp_num(self, pool=None):
+            k = 0
+            m = 99
+            has_pools = False
+            pools_json = self.get_osd_dump_json()['pools']
```
```
, u'epoch': 1, u'modified': u'0.000000'}, u'fsid': u'9f9bb9d1-8011-4afc-bc3a-237e70dc3946', u'quorum': [0, 1, 2]}
2019-04-03T07:55:22.282 INFO:tasks.thrashosds.thrasher:Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2019-04-01-1531/qa/tasks/ceph_manager.py", line 953, in wrapper
    return func(self)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2019-04-01-1531/qa/tasks/ceph_manager.py", line 1083, in do_thrash
    self.choose_action()()
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2019-04-01-1531/qa/tasks/ceph_manager.py", line 703, in test_pool_min_size
    pools_json = self.get_osd_dump_json()['pools']
AttributeError: Thrasher instance has no attribute 'get_osd_dump_json'
```

http://pulpito.ceph.com/kchai-2019-04-03_07:35:34-rados-wip-kefu-testing-2019-04-01-1531-distro-basic-smithi/3804803

`get_osd_dump_json()` is a method of `CephManager` not `Thrasher`.

Liuchang0812 commented:
updated `qa/tasks/ceph_manager.py`



http://pulpito.ceph.com/kchai-2019-04-01_10:38:29-rados-wip-kefu-testing-2019-04-01-1531-distro-basic-mira/3797065/ 

I created an issue to track this failure: http://tracker.ceph.com/issues/39087.

tchaikov commented on qa/tasks/ceph_manager.py:
```diff
@@ -637,7 +638,6 @@ def grow_pool(self):
         Increase the size of the pool
         """
         pool = self.ceph_manager.get_pool()
-        orig_pg_num = self.ceph_manager.get_pool_pg_num(pool)
```
why are you removing this line?

Liuchang0812 replied:
```
➜  ceph git:(wip-ec-below-min-size) ✗ grep -nri "orig_pg_num" qa/tasks/ceph_manager.py
➜  ceph git:(wip-ec-below-min-size) ✗ 
```
it seems `orig_pg_num` is a variable not used.

tchaikov replied:
ahh, i see. i forgot to remove it in c0f0cde399a80cea617c115417e5390a63ec674e .

thanks!

neha-ojha commented:
> updated `qa/tasks/ceph_manager.py`
> 
> http://pulpito.ceph.com/kchai-2019-04-01_10:38:29-rados-wip-kefu-testing-2019-04-01-1531-distro-basic-mira/3797065/
> 
> I created an issue to track this failure: http://tracker.ceph.com/issues/39087.

@xiexingguo it looks like "num_missing": 3 and "num_unfound": 3 are same here, do you think https://github.com/ceph/ceph/pull/27205 could have fixed this?

xiexingguo commented:
> @xiexingguo it looks like "num_missing": 3 and "num_unfound": 3 are same here, do you think #27205 could have fixed this?

@neha-ojha 

No, it's quite a different bug.
It seems primary (**2.4s0**) is trying to collect OI of "2:29be9562:::dummy:head" from shard pg 2.4s1 during scrubbing, but **2.4s1** is being unable to find that object (**collection_list returns 0 objects**):

```
2019-04-01 15:40:40.622 7feae266b700 10 osd.0 pg_epoch: 62 pg[2.4s1( v 30'5134 (30'2100,30'5134] local-lis/les=41/42 n=4 ec=15/15 lis/c 41/41 les/c/f 42/42/0 28/41/41)
 [1,0,2,3]p1(0) r=1 lpr=41 luod=0'0 crt=30'5134 lcod 30'5133 active mbc={}] build_scrub_map_chunk [2:20000000::::0,MAX)  pos (0/0)
2019-04-01 15:40:40.622 7feae266b700 15 bluestore(/var/lib/ceph/osd/ceph-0) collection_list 2.4s1_head start 1#2:20000000::::0# end 1#MAX# max 2147483647
2019-04-01 15:40:40.622 7feae266b700 20 bluestore(/var/lib/ceph/osd/ceph-0) _collection_list range 0x817ffffffffffffffc20000000 to 0x817ffffffffffffffc30000000 and 0x8
1800000000000000220000000 to 0x81800000000000000230000000 start 1#2:20000000::::0#
2019-04-01 15:40:40.622 7feae266b700 20 bluestore(/var/lib/ceph/osd/ceph-0) _collection_list start from 0x8180000000000000022000000021213d0000000000000000fffffffffffff
fff'o' temp=0
2019-04-01 15:40:40.622 7feae266b700 10 bluefs _read_random h 0x55f5b34788c0 0x304511~1072 from file(ino 38 size 0x2108874 mtime 2019-04-01 14:52:13.346917 bdev 1 allo
cated 2200000 extents [1:0xaccc00000~400000,1:0xacdd00000~400000,1:0xacee00000~400000,1:0xacfe00000~400000,1:0xad1000000~400000,1:0xad2100000~400000,1:0xad3200000~4000
00,1:0xad4500000~400000,1:0xad5a00000~200000])
2019-04-01 15:40:40.622 7feae266b700 20 bluefs _read_random read buffered 0x304511~1072 of 1:0xaccc00000~400000
2019-04-01 15:40:40.626 7feafef30700  1 -- v2:172.21.9.136:6804/43475 <== osd.2 172.21.9.136:0/37794 7 ==== osd_ping(ping e62 stamp 2019-04-01 15:40:40.626641) v4 ====
 2004+0+0 (crc 0 0 0) 0x55f5b1fbf680 con 0x55f59e3ad800
2019-04-01 15:40:40.626 7feafef30700  1 -- v2:172.21.9.136:6804/43475 --> 172.21.9.136:0/37794 -- osd_ping(ping_reply e62 stamp 2019-04-01 15:40:40.626641) v4 -- 0x55f
5b1fbfc20 con 0x55f59e3ad800
2019-04-01 15:40:40.626 7feafef30700 20 osd.0 62 share_map_peer 0x55f59d70c000 already has epoch 62
2019-04-01 15:40:40.626 7feaff731700  1 -- v2:172.21.9.136:6803/43475 <== osd.2 172.21.9.136:0/37794 7 ==== osd_ping(ping e62 stamp 2019-04-01 15:40:40.626641) v4 ====
 2004+0+0 (crc 0 0 0) 0x55f5bcae3860 con 0x55f59e51d000
2019-04-01 15:40:40.626 7feaff731700  1 -- v2:172.21.9.136:6803/43475 --> 172.21.9.136:0/37794 -- osd_ping(ping_reply e62 stamp 2019-04-01 15:40:40.626641) v4 -- 0x55f
59e2314a0 con 0x55f59e51d000
2019-04-01 15:40:40.626 7feaff731700 20 osd.0 62 share_map_peer 0x55f59d70c000 already has epoch 62
2019-04-01 15:40:40.637 7feae266b700 20 bluefs _read_random got 4210
2019-04-01 15:40:40.637 7feae266b700 20 bluestore(/var/lib/ceph/osd/ceph-0) _collection_list pend 0x81800000000000000230000000
2019-04-01 15:40:40.637 7feae266b700 20 bluestore(/var/lib/ceph/osd/ceph-0) _collection_list iterator not valid (end of db?)
2019-04-01 15:40:40.637 7feae266b700 10 bluestore(/var/lib/ceph/osd/ceph-0) collection_list 2.4s1_head start 1#2:20000000::::0# end 1#MAX# max 2147483647 = 0, ls.size(
) = 0, next = GHMIN
```

But that object **should** exist on 2.4s1, e.g., **because** fsck is able to read from that object correctly early on boot:

```
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _fsck  collection 2.4s1_head cnode(bits 4)
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _fsck  collection 2.4s1_head cnode(bits 4)
2019-04-01 14:49:17.210 7feb05df4d80 10 bluestore(/var/lib/ceph/osd/ceph-0) _fsck  1#2:20000000::::head#
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0).collection(2.4s1_head 0x55f59d1ceb40) get_onode oid 1#2:20000000::::head# key 0x81800000000
00000022000000021213dfffffffffffffffeffffffffffffffff'o'
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0).collection(2.4s1_head 0x55f59d1ceb40)  r 0 v.len 29
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read 0x0~0 size 0x0 (0)
2019-04-01 14:49:17.210 7feb05df4d80 10 bluestore(/var/lib/ceph/osd/ceph-0) _fsck  1#2:227e0c77:::existing_7:head#
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0).collection(2.4s1_head 0x55f59d1ceb40) get_onode oid 1#2:227e0c77:::existing_7:head# key 0x8
18000000000000002227e0c'w!existing_7!='0xfffffffffffffffeffffffffffffffff'o'
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0).collection(2.4s1_head 0x55f59d1ceb40)  r 0 v.len 483
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore.blob(0x55f59d584ee0) get_ref 0x0~1000 Blob(0x55f59d584ee0 blob([0x340000~10000] csum+has_unused crc32c/0x1000 unused=
0xfffe) use_tracker(0x0 0x0) SharedBlob(0x55f59d584f50 sbid 0x0))
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore.blob(0x55f59d584ee0) get_ref init 0x10000, 10000
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _fsck    0x0~1000: 0x0~1000 Blob(0x55f59d584ee0 blob([0x340000~10000] csum+has_unused crc32
c/0x1000 unused=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d584f50 sbid 0x0))
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _fsck  referenced 0x1 for Blob(0x55f59d584ee0 blob([0x340000~10000] csum+has_unused crc32c/
0x1000 unused=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d584f50 sbid 0x0))
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read 0x0~1000 size 0x1000 (4096)
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read  blob Blob(0x55f59d584ee0 blob([0x340000~10000] csum+has_unused crc32c/0x1000 unus
ed=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d584f50 sbid 0x0)) need 0x0~1000 cache has 0x[]
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read  blob Blob(0x55f59d584ee0 blob([0x340000~10000] csum+has_unused crc32c/0x1000 unus
ed=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d584f50 sbid 0x0)) need {<0x0, 0x1000> : [0x0:0~1000]}
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read    region 0x0: 0x0 reading 0x0~1000
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read  blob Blob(0x55f59d584ee0 blob([0x340000~10000] csum+has_unused crc32c/0x1000 unus
ed=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d584f50 sbid 0x0)) need 0x{<0x0, 0x1000> : [0x0:0~1000]}
2019-04-01 14:49:17.210 7feb05df4d80 10 bluestore(/var/lib/ceph/osd/ceph-0) _fsck  1#2:29be9562:::dummy:head#
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0).collection(2.4s1_head 0x55f59d1ceb40) get_onode oid 1#2:29be9562:::dummy:head# key 0x818000
00000000000229be95'b!dummy!='0xfffffffffffffffeffffffffffffffff'o'
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0).collection(2.4s1_head 0x55f59d1ceb40)  r 0 v.len 478
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore.blob(0x55f59d584fc0) get_ref 0x0~1000 Blob(0x55f59d584fc0 blob([0x270000~10000] csum+has_unused crc32c/0x1000 unused=
0xfffe) use_tracker(0x0 0x0) SharedBlob(0x55f59d585030 sbid 0x0))
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore.blob(0x55f59d584fc0) get_ref init 0x10000, 10000
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _fsck    0x0~1000: 0x0~1000 Blob(0x55f59d584fc0 blob([0x270000~10000] csum+has_unused crc32
c/0x1000 unused=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d585030 sbid 0x0))
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _fsck  referenced 0x1 for Blob(0x55f59d584fc0 blob([0x270000~10000] csum+has_unused crc32c/
0x1000 unused=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d585030 sbid 0x0))
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read 0x0~1000 size 0x1000 (4096)
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read  blob Blob(0x55f59d584fc0 blob([0x270000~10000] csum+has_unused crc32c/0x1000 unus
ed=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d585030 sbid 0x0)) need 0x0~1000 cache has 0x[]
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read  blob Blob(0x55f59d584fc0 blob([0x270000~10000] csum+has_unused crc32c/0x1000 unus
ed=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d585030 sbid 0x0)) need {<0x0, 0x1000> : [0x0:0~1000]}
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read    region 0x0: 0x0 reading 0x0~1000
2019-04-01 14:49:17.210 7feb05df4d80 20 bluestore(/var/lib/ceph/osd/ceph-0) _do_read  blob Blob(0x55f59d584fc0 blob([0x270000~10000] csum+has_unused crc32c/0x1000 unus
ed=0xfffe) use_tracker(0x10000 0x1000) SharedBlob(0x55f59d585030 sbid 0x0)) need 0x{<0x0, 0x1000> : [0x0:0~1000]}
```

and in between I failed to find any evidence that object is deleted.

liewegas commented:
```
2019-04-05T08:03:08.136 INFO:tasks.thrashosds.thrasher:choosing based on number of live OSDs and rand val 0.662706817037
2019-04-05T08:03:08.137 INFO:tasks.thrashosds.thrasher:Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 953, in wrapper
    return func(self)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 1083, in do_thrash
    self.choose_action()()
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 752, in test_pool_min_size
    if len(self.live_osds) > minup+1 and rand_val < 0.5:
TypeError: coercing to Unicode: need string or buffer, int found

2019-04-05T08:03:08.138 ERROR:teuthology:Uncaught exception (Hub)
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/greenlet.py", line 536, in run
    result = self._run(*self.args, **self.kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 953, in wrapper
    return func(self)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 1083, in do_thrash
    self.choose_action()()
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 752, in test_pool_min_size
    if len(self.live_osds) > minup+1 and rand_val < 0.5:
TypeError: coercing to Unicode: need string or buffer, int found
```

Liuchang0812 commented:
> ```
> 2019-04-05T08:03:08.136 INFO:tasks.thrashosds.thrasher:choosing based on number of live OSDs and rand val 0.662706817037
> 2019-04-05T08:03:08.137 INFO:tasks.thrashosds.thrasher:Traceback (most recent call last):
>   File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 953, in wrapper
>     return func(self)
>   File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 1083, in do_thrash
>     self.choose_action()()
>   File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 752, in test_pool_min_size
>     if len(self.live_osds) > minup+1 and rand_val < 0.5:
> TypeError: coercing to Unicode: need string or buffer, int found
> 
> 2019-04-05T08:03:08.138 ERROR:teuthology:Uncaught exception (Hub)
> Traceback (most recent call last):
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/gevent/greenlet.py", line 536, in run
>     result = self._run(*self.args, **self.kwargs)
>   File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 953, in wrapper
>     return func(self)
>   File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 1083, in do_thrash
>     self.choose_action()()
>   File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-sage-testing-2019-04-03-0955/qa/tasks/ceph_manager.py", line 752, in test_pool_min_size
>     if len(self.live_osds) > minup+1 and rand_val < 0.5:
> TypeError: coercing to Unicode: need string or buffer, int found
> ```
> /a/sage-2019-04-03_19:24:47-rados-wip-sage-testing-2019-04-03-0955-distro-basic-smithi/3805336

thank you @liewegas , I updated ceph_manager.py. 

btw,  how do you test the teuthology code? I cant not run-and-test those python code locally. any suggestions?

tchaikov commented:
@Liuchang0812 i think you need to either setup a teuthology instance (and other supporting service) for running this teuthology based tests. or you can apply for accessing sepia lab for running them in sepia. see https://ceph.github.io/sepia/adding_users/

Liuchang0812 commented:
thank you, Kefu. I created an issue: http://tracker.ceph.com/issues/39153

Kefu Chai <notifications@github.com> 于2019年4月9日周二 下午2:00写道：

> @Liuchang0812 <https://github.com/Liuchang0812> i think you need to
> either setup a teuthology instance (and other supporting service) for
> running this teuthology based tests. or you can apply for accessing sepia
> lab for running them in sepia. see
> https://ceph.github.io/sepia/adding_users/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/17619#issuecomment-481113613>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABtHdjVFguP8gz-k9wCkikd2w8zgSpwiks5vfCxygaJpZM4PSUOc>
> .
>

liewegas commented:
in theory you can add the targets: yaml section with available machiens to 
a yaml file with your job and run teuthology directly from your git 
checkout with 'virtualenv/bin/teuthology foo.yaml' or similar.  I haven't 
done this in a while and there are generally some things to fiddle with to 
get it to work but it can be done.

The other way is to pass --teuthology-branch to teuthology-queue or 
teuthology-suite.

tchaikov commented:
- http://pulpito.ceph.com/kchai-2019-04-09_12:51:48-rados-wip-kefu-testing-2019-04-09-1348-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-04-11_01:06:08-rados-wip-kefu-testing-2019-04-09-1348-distro-basic-smithi/

the only genuine failure is https://tracker.ceph.com/issues/23947

retest this please.

neha-ojha commented:
These failures in the first run might be worth investigating.
http://pulpito.ceph.com/kchai-2019-04-09_12:51:48-rados-wip-kefu-testing-2019-04-09-1348-distro-basic-mira/3828057/ - failed to become clean before timeout expired on an ec test
http://pulpito.ceph.com/kchai-2019-04-09_12:51:48-rados-wip-kefu-testing-2019-04-09-1348-distro-basic-mira/3828182/

@gregsfortytwo want to take another look at this PR?

Liuchang0812 commented:
> These failures in the first run might be worth investigating.
> http://pulpito.ceph.com/kchai-2019-04-09_12:51:48-rados-wip-kefu-testing-2019-04-09-1348-distro-basic-mira/3828057/ - failed to become clean before timeout expired on an ec test
> http://pulpito.ceph.com/kchai-2019-04-09_12:51:48-rados-wip-kefu-testing-2019-04-09-1348-distro-basic-mira/3828182/
> 
> @gregsfortytwo want to take another look at this PR?

```
2019-04-10T08:04:25.698 INFO:tasks.ceph.osd.2.mira018.stderr:2019-04-10 08:03:37.122 7f75570db700 -1 bdev(0x55f61268b180 /var/lib/ceph/osd/ceph-2/block) _aio_thread stalled aio 0x55f6132a8150 since 2019-04-10 08:02:37.092398, timeout is 60s, suicide                                                         
2019-04-10T08:04:25.698 INFO:tasks.ceph.osd.2.mira018.stderr:2019-04-10 08:03:37.123 7f75570db700 -1 /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/15.0.0-234-g1153fd1/rpm/el7/BUILD/ceph-15.0.0-234-g1153fd1/src/os/bluestore/KernelDevice.cc: In function 'void KernelDevice::_aio_thread()' thread 7f75570db700 time 2019-04-10 08:03:37.123459
2019-04-10T08:04:25.698 INFO:tasks.ceph.osd.2.mira018.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/15.0.0-234-g1153fd1/rpm/el7/BUILD/ceph-15.0.0-234-g1153fd1/src/os/bluestore/KernelDevice.cc: 574: abort()
2019-04-10T08:04:25.698 INFO:tasks.ceph.osd.2.mira018.stderr:                                         
2019-04-10T08:04:25.699 INFO:tasks.ceph.osd.2.mira018.stderr: ceph version 15.0.0-234-g1153fd1 (1153fd1988712e15c4ad07e04fb3604e4eafde7e) octopus (dev)
2019-04-10T08:04:25.699 INFO:tasks.ceph.osd.2.mira018.stderr: 1: (ceph::__ceph_abort(char const*, int, char const*, std::string const&)+0xd8) [0x55f60656a0d5]
2019-04-10T08:04:25.699 INFO:tasks.ceph.osd.2.mira018.stderr: 2: (KernelDevice::_aio_thread()+0x1008) [0x55f606c714b8]
2019-04-10T08:04:25.699 INFO:tasks.ceph.osd.2.mira018.stderr: 3: (KernelDevice::AioCompletionThread::entry()+0xd) [0x55f606c779fd]
2019-04-10T08:04:25.700 INFO:tasks.ceph.osd.2.mira018.stderr: 4: (()+0x7dd5) [0x7f7563486dd5]         
2019-04-10T08:04:25.700 INFO:tasks.ceph.osd.2.mira018.stderr: 5: (clone()+0x6d) [0x7f756234cead]      
```
an OSD suicided

gregsfortytwo commented:
The thrasher looks fine. I'd like us to quantify how often we're actually forcing it to do recovery below min_size before we merge though, to make sure it's actually occurring!

gregsfortytwo commented:
And thanks for sticking with this @Liuchang0812!

gregsfortytwo commented:
Jenkins retest this please

neha-ojha commented:
@Liuchang0812 could you please rebase

Liuchang0812 commented:
> @Liuchang0812 could you please rebase

done

neha-ojha commented:
> > @Liuchang0812 could you please rebase
> 
> ~done~ sorry...I think I need more time to do it.

sure, no rush.

Liuchang0812 commented:
> > > @Liuchang0812 could you please rebase
> > 
> > 
> > ~done~ sorry...I think I need more time to do it.
> 
> sure, no rush.

would you mind running a teuthology test again?

neha-ojha commented:
@tchaikov are you already testing this PR? unfortunately, my last run did not finish successfully

tchaikov commented:
@neha-ojha my run has not finished yet. see http://pulpito.ceph.com/kchai-2019-06-02_15:00:54-rados-wip-kefu-testing-2019-06-01-2346-distro-basic-mira/ 

and i have not checked the failures yet. i planned to rerun the failed tests after the qa run finished.

neha-ojha commented:
> @neha-ojha my run has not finished yet. see http://pulpito.ceph.com/kchai-2019-06-02_15:00:54-rados-wip-kefu-testing-2019-06-01-2346-distro-basic-mira/
> 
> and i have not checked the failures yet. i planned to rerun the failed tests after the qa run finished.

I have scheduled the rerun: http://pulpito.ceph.com/nojha-2019-06-04_22:13:54-rados-wip-kefu-testing-2019-06-01-2346-distro-basic-smithi/

tchaikov commented:
- http://pulpito.ceph.com/kchai-2019-06-02_15:00:54-rados-wip-kefu-testing-2019-06-01-2346-distro-basic-mira/
- http://pulpito.ceph.com/nojha-2019-06-04_22:13:54-rados-wip-kefu-testing-2019-06-01-2346-distro-basic-smithi/

the valgrind failure is a known issue tracked by http://tracker.ceph.com/issues/38403, and it should be addressed by #28397

tchaikov commented:
@gregsfortytwo did the latest change address your comment?

@neha-ojha thanks for scheduling the rerun! i think the qa runs are sane.

dzafman commented on src/osd/PeeringState.cc:
```diff
@@ -1607,22 +1607,21 @@ bool PeeringState::recoverable_and_ge_min_size(const vector<int> &want) const
+  if (num_want_acting < pool.info.min_size) {
+    if (cct->_conf->osd_allow_recovery_below_min_size && 
+	missing_loc.get_recoverable_predicate()(have)) {
+      dout(20) << "choose_acting successfully, we allow osd recovery below min "
```
successful?

neha-ojha replied:
can we use `psdout` to be consistent? and also using `<< __func__ <<` instead of "choose_acting" might be better, since this function is called from more than one place.

dzafman commented on src/osd/PeeringState.cc:
```diff
@@ -1607,22 +1607,21 @@ bool PeeringState::recoverable_and_ge_min_size(const vector<int> &want) const
           pool.info.is_erasure() ? shard_id_t(i) : shard_id_t::NO_SHARD));
     }
   }
-  // We go incomplete if below min_size for ec_pools since backfill
```
Should you update recoverable_and_ge_min_size()?

liuchang0812 replied:
thank you, @dzafman, how about `PeeringState::recoverable` ?

dzafman replied:
sure

neha-ojha commented on src/osd/PeeringState.cc:
```diff
@@ -1607,22 +1607,21 @@ bool PeeringState::recoverable_and_ge_min_size(const vector<int> &want) const
+               << dendl;
+      // do nothing here
+    } else {
+      dout(10) << "choose_acting failed, below min size and is not recoverable "
```
ditto

gregsfortytwo commented on qa/tasks/ceph_manager.py:
```diff
@@ -149,6 +149,7 @@ def tmp(x):
         # another
         first_mon = teuthology.get_first_mon(manager.ctx, self.config).split('.')
         opts = [('mon', 'mon_osd_down_out_interval', 0)]
+        #why do we disable marking an OSD out automatically? :/
```
Did this question get dealt with?

gregsfortytwo commented:
> @gregsfortytwo did the latest change address your comment?

I don't see anything checking in the source to try and count recovery-from-below-min-size, did someone examine the logs to show it happening a lot?
(I see the tester is pretty aggressive about marking down a whole lot of OSDs at once so I imagine the odds are quite good now, anyway!)

neha-ojha commented:
> > @gregsfortytwo did the latest change address your comment?
> 
> I don't see anything checking in the source to try and count recovery-from-below-min-size, did someone examine the logs to show it happening a lot?
> (I see the tester is pretty aggressive about marking down a whole lot of OSDs at once so I imagine the odds are quite good now, anyway!)

@gregsfortytwo I checked the latest test results from @tchaikov and verified that it is happening.

neha-ojha commented on src/osd/PeeringState.cc:
```diff
@@ -1607,22 +1607,21 @@ bool PeeringState::recoverable_and_ge_min_size(const vector<int> &want) const
-  if (!missing_loc.get_recoverable_predicate()(have)) {
-    psdout(10) << __func__ << " failed, not recoverable" << dendl;
-    return false;
+  if (num_want_acting < pool.info.min_size) {
```
@liuchang0812 sorry for not bringing this up earlier, but I think we want to check for `SERVER_OCTOPUS` here to use the new code path and we should revive the old code if this condition does not hold true. This is because in mixed cluster scenarios(for example nautilus and octopus), inconsistent `choose_acting()` behavior can lead to a loop, changing pg_tmp back and forth.

liewegas replied:
alternatively, we might want to condition this on the OSDMap's require_osd_release. Is there potential for problems if we have a smaller acting set with octopus features, and then later a larger/different acting set that doesn't?

neha-ojha replied:
I'm not sure, do you see a problem in using `SERVER_OCTOPUS`? either way, we should avoid getting into a situation like https://tracker.ceph.com/issues/39441

liewegas replied:
I'm wondering if it's possible to get into a situation where there is disagreement in the PastIntervals as far as whether a particular interval is maybe_rw or not... if differnet OSDs disagreed about that it could cause some weirdness.  But I think this PR isn't changing the active criteria?  In which case I think it's fine...

tchaikov commented on src/osd/PeeringState.cc:
```diff
@@ -1594,36 +1594,68 @@ void PeeringState::calc_replicated_acting(
-  }
+bool PeeringState::recoverable(const vector<int> &want) const
+{
+
```
please remove this empty line.

tchaikov commented on src/osd/PeeringState.cc:
```diff
@@ -1594,36 +1594,68 @@ void PeeringState::calc_replicated_acting(
-    return false;
-  }
+    /* Check whether we have enough acting shards to later perform recovery */
+    if (num_want_acting < pool.info.min_size) {
```
shall we still check `missing_loc.get_recoverable_predicate()(have)` even if `num_want_acting >= pool.info.min_size` ?

also, i'd suggest consolidate these checks to improve the readability:
```c++
if (num_want_acting < pool.info.min_size) {
  const bool recovery_ec_pg_under_min_size =
    HAVE_FEATURE(get_osdmap()->get_up_osd_features(), SERVER_OCTOPUS);
  if (pool.info.is_erasure() && !recovery_ec_pg_under_min_size) {
    psdout(10) << __func__ << " failed, ec recovery below min size not supported pre-preoctopus" << dendl;
    return false;
  } else if (!cct->_conf->osd_allow_recovery_below_min_size) {
    psdout(10) << __func__ << " failed, recovery below min size not enabled" << dendl;
    return false;
  }
}
if (missing_loc.get_recoverable_predicate()(have)) {
  return true;
} else {
  psdout(10) << __func__ << " failed, not recoverable " << dendl;
  return false;
}
```

liuchang0812 replied:
thank you, @tchaikov . 

> shall we still check missing_loc.get_recoverable_predicate()(have) even if num_want_acting >= pool.info.min_size ?

yes, we should check it. i wonder if user set wrong min size on ec pool. ec pool is not recoverable even `num_want_acting >= pool.info.min` ( k=3 m=2 min_size=2 num_want_acting=2)

tchaikov commented on src/osd/PeeringState.cc:
```diff
@@ -1611,10 +1611,11 @@ bool PeeringState::recoverable(const vector<int> &want) const
+
     if (pool.info.is_erasure() && !recovery_ec_pg_under_min_size) {
-      psdout(10) << __func__ << " failed, ec recovery below min size not supported pre-preoctopus" << dendl;
+      psdout(10) << __func__ << " failed, ec recovery below min size not supported pre-octopus" << dendl;
```
s/supported/supported by/

liuchang0812 commented:
ping @neha-ojha

neha-ojha commented:
rados: http://pulpito.ceph.com/nojha-2019-06-17_20:20:02-rados-wip-ec-below-min-size-2019-06-17-distro-basic-smithi/

@liuchang0812 http://pulpito.ceph.com/nojha-2019-06-17_20:20:02-rados-wip-ec-below-min-size-2019-06-17-distro-basic-smithi/4043810/ this failure uses minsize_recovery.yaml, can you take a look?

upgrade: http://pulpito.ceph.com/nojha-2019-06-17_20:26:58-upgrade-wip-ec-below-min-size-2019-06-17-distro-basic-smithi/ - This one saw a lot of failures due to the same root cause. This issue has now been fixed, so I'll rerun the upgrade tests.

neha-ojha commented:
rados rerun: http://pulpito.ceph.com/nojha-2019-06-20_16:59:16-rados-wip-ec-below-min-size-2019-06-17-distro-basic-smithi/ - no related failures

new upgrade run: http://pulpito.ceph.com/nojha-2019-06-20_17:02:47-upgrade-wip-ec-below-min-size-2019-06-17-distro-basic-smithi/ - no related failures

gregsfortytwo commented:
I’ve been looking forward to getting this in; let’s do it!

On Fri, Jun 21, 2019 at 4:36 PM Neha Ojha <notifications@github.com> wrote:

> *@neha-ojha* approved this pull request.
>
> @gregsfortytwo <https://github.com/gregsfortytwo> @tchaikov
> <https://github.com/tchaikov> looks good to go to me, thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/17619?email_source=notifications&email_token=AAG5ZP5AFDI3NHY7QXUEKGDP3VQVVA5CNFSM4D2JIOOKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4KX5EI#pullrequestreview-253066897>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAG5ZP4ZTJN4J2USN63T2TTP3VQVVANCNFSM4D2JIOOA>
> .
>

neha-ojha commented:
@gregsfortytwo you had requested changes, if everything has been addressed, can you remove it?

hoerup commented:
Will this be backportet ?

neha-ojha commented:
> Will this be backportet ?

We do not have plans of backporting this.

hoerup commented:
@neha-ojha thanks for the answer. I have updated https://tracker.ceph.com/issues/40488 accordingly
